### ISSUE TITLE
Include line numbers in jinja/yaml error messages

### ISSUE THREAD

**Issue by viraptor** [saltstack/salt#21298]
Currently (in 2014.7.0), if there's some parameter missing when rendering a template, I get a not very helpful message, like:

```
          ID: /file/path
    Function: file.managed
      Result: False
     Comment: Unable to manage file: Jinja variable 'dict object' has no attribute 'user'
     Started: 11:47:12.728811
    Duration: 37.236 ms
     Changes:   
```

It would be great if the line number from the template was also included, if jinja itself gives this information. Otherwise it can be quite hard to find where the issue occurred.

**Comment by eliasp** [comment]
I agree. Something I'd like to see as well.

**Comment by jfindlay** [comment]
@viraptor, I've changed the title to specifically refer to the line number issue, which is an excellent idea.  Thanks.

**Comment by uvsmtid** [comment]
+1
Related #4236.

### Reiterating the problem ###

Yes, it's frustrating when I receive this error for a long template (like those generating some XML config file):
```
SaltRenderError: Jinja variable 'dict object' has no attribute 'None'
```
* Which `'dict object'` among all nested `pillar[subkey1][subkey2]...`?
* Where did the key `'None'` come from?

### Searching for improvements ###

I understand that it is largely the issue of Jinja, but any fix/improvement/advice to reduce scope of entire SLS file to fewer lines is acceptable:
* line numbers;
* actual variable name (not `'dict object'` or `attribute 'user'`);
* contents of the 'dict object' (which indicate candidates among all possible dict objects template may use).

### Current workaround ###

At the moment I only know one approach: go and manually disable portions of the templates by `{% if False %}` / `{% endif %}` (or Jinja comments) until the problem is spotted.

I did a quick (unsuccessful) search for any way to put stderr debug output by Jinja (like printing contents of some `'dict object'`) which would be seen when running `salt-call`. That would help.

**Comment by gravyboat** [comment]
+1

**Comment by chris-martin** [comment]
I just templated a config file and the result is `Jinja variable 'dict object' has no attribute 'staging'`. There are 30+ lines in my file that that error could apply to. I'm a very sad panda right now :crying_cat_face:

**Comment by chris-martin** [comment]
Since it sounds like this is jinja's fault, is there a better templating library supported by salt that does give decent error output?

**Comment by jfindlay** [comment]
@chris-martin, there are a few 'renderers' that are supported by salt.  You can try the mako or py renderers.

**Comment by The-Loeki** [comment]
#37317 & me still care ;)

**Comment by viraptor** [comment]
Don't touch it, bit...

**Comment by sagetherage** [comment]
the linked PR needs a test case written and then the release cycle can be determined

**Comment by sagetherage** [comment]
@waynew another good one for the Test Clinic, assigning you only in attempt to help write tests. We will still need tests written before we can commit to a release cycle for this issue and PR, but thank you all for your contributions!

**Comment by whytewolf** [comment]
so jinja now does point out as close as it can where the error is it isn't aways perfect because logic can be tricky. but at least should get to a general location for most errors.

see following example 

```
root@salt00:/srv/salt/tests# salt salt00 slsutil.renderer salt://tests/fail.sls default_renderer=jinja
salt00:
    The minion function caused an exception: Traceback (most recent call last):
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/templates.py", line 476, in render_jinja_tmpl
        output = template.render(**decoded_context)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/jinja2/environment.py", line 1301, in render
        self.environment.handle_exception()
      File "/opt/saltstack/salt/lib/python3.10/site-packages/jinja2/environment.py", line 936, in handle_exception
        raise rewrite_traceback_stack(source=source)
      File "<template>", line 1, in top-level template code
    jinja2.exceptions.UndefinedError: 'nope' is undefined

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 1982, in _thread_return
        return_data = minion_instance._execute_job_function(
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 1941, in _execute_job_function
        return_data = self.executors[fname](opts, data, func, args, kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 159, in __call__
        ret = self.loader.run(run_func, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1245, in run
        return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1260, in _run_as
        return _func_or_method(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/executors/direct_call.py", line 10, in execute
        return func(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 159, in __call__
        ret = self.loader.run(run_func, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1245, in run
        return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1260, in _run_as
        return _func_or_method(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/slsutil.py", line 176, in renderer
        ret = salt.template.compile_template(
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/template.py", line 99, in compile_template
        ret = render(input_data, saltenv, sls, **render_kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 159, in __call__
        ret = self.loader.run(run_func, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1245, in run
        return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1260, in _run_as
        return _func_or_method(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/renderers/jinja.py", line 62, in render
        tmp_data = salt.utils.templates.JINJA(
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/templates.py", line 218, in render_tmpl
        output = render_str(tmplstr, context, tmplpath)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/utils/templates.py", line 482, in render_jinja_tmpl
        raise SaltRenderError("Jinja variable {}{}".format(exc, out), line, tmplstr)
    salt.exceptions.SaltRenderError: Jinja variable 'nope' is undefined; line 1

    ---
    {{nope}}    <======================
    ---
```

I'm going to close this as complete for now. if you try the latest and are not getting some kind of feed back it would be a bug. and a new ticket should be made in that vain.
